### PR TITLE
Bug fixes

### LINKED ISSUE #3761: Live markdown fails with NRE
Repro markdown:
https://raw.githubusercontent.com/Microsoft/acceleratoRs/master/CreditRiskPrediction/Code/CreditRiskDeploy.Rmd

image

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
#3761 Live markdown fails with NRE
- Check that lines collection is not null

#3818 Intellisense doesn't like ! in arithmetic expr
- Special case `!` since it applies to the entire expression that follows and is not like C unary `!`
- Update tests

#3767 Closing VS throws: Connection point was not released
- Dispose language settings storage on package unload

#3816 IntelliSense not prompt global variable in [ ]
- Make sure we do include global variables in completion inside indexer

#3814 Annoying behavior with Rmd files when code is in last line
- Preserve viewport position along with the caret position
- Do not replace spans if they are identical

Also, exclude templates from VSIX since they are in the extension manifest.

### CODE REVIEW

File: src/R/Core/Impl/AST/Expressions/ExpressionParser.cs [hunk 1]
```diff
@@ -383,7 +383,7 @@ enum OperationType {
         }
 
         private OperationType HandleOperator(ParseContext context, out ParseErrorType errorType) {
-            errorType = this.HandleOperator(context, null, out bool isUnary);
+            errorType = this.HandleOperator(context, out bool isUnary);
```
Comment (AlexanderSher) [inline src/R/Core/Impl/AST/Expressions/ExpressionParser.cs]: You don't need `this`

Comment (MikhailArkhipov) [inline src/R/Core/Impl/AST/Expressions/ExpressionParser.cs]: Lol I've read I don't need this `code` ;-)

File: src/Windows/Editor/Impl/Projection/ProjectionBufferManager.cs [hunk 2]
```diff
@@ -17,7 +18,12 @@ namespace Microsoft.Languages.Editor.Projection {
     /// Manages the projection buffer for the primary language
     /// </summary>
     public sealed class ProjectionBufferManager : IProjectionBufferManager {
-        private int? _savedCaretPosition;
+        private class ViewPosition {
```
Comment (AlexanderSher) [inline src/Windows/Editor/Impl/Projection/ProjectionBufferManager.cs]: you can use tuples for this

Comment (MikhailArkhipov) [inline src/Windows/Editor/Impl/Projection/ProjectionBufferManager.cs]: Kind of looks cleaner.

File: src/Windows/Editor/Impl/Projection/ProjectionBufferManager.cs [hunk 3]
```diff
@@ -151,17 +161,68 @@ public sealed class ProjectionBufferManager : IProjectionBufferManager {
...
+            _savedViewPosition = new ViewPosition {
+                CaretPosition = GetCaretPosition(),
+                ViewportTop = DiskBuffer.GetFirstView()?.ViewportTop
+            };
+        }
+
+        private void RestoreViewPosition() {
+            var textView = DiskBuffer.GetFirstView();
+            if(textView == null) {
+                return;
```
Comment (AlexanderSher) [inline src/Windows/Editor/Impl/Projection/ProjectionBufferManager.cs]: Shouldn't we set `_savedViewPosition = null;` here as well?

**Comment by karthiknadig** [comment]
:shipit:

File: src/Windows/Editor/Impl/Projection/ProjectionBufferManager.cs (hunk 3 above)
Comment (MikhailArkhipov) [inline src/Windows/Editor/Impl/Projection/ProjectionBufferManager.cs]: OK. Generally this is an odd case though.

**Comment by AlexanderSher** [comment]
👍

**Pull request merged**